modules/table/src/main/java/org/apache/ignite/internal/table/distributed/CatalogStorageIndexDescriptorSupplier.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed;
+
+import org.apache.ignite.internal.catalog.CatalogService;
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexDescriptor;
+import org.apache.ignite.internal.catalog.descriptors.CatalogTableDescriptor;
+import org.apache.ignite.internal.hlc.HybridTimestamp;
+import org.apache.ignite.internal.storage.index.StorageIndexDescriptor;
+import org.apache.ignite.internal.storage.index.StorageIndexDescriptorSupplier;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Index descriptor supplier from catalog.
+ */
+class CatalogStorageIndexDescriptorSupplier implements StorageIndexDescriptorSupplier {
+    private final CatalogService catalogService;
+
+    private final LowWatermark lowWatermark;
+
+    CatalogStorageIndexDescriptorSupplier(CatalogService catalogService, LowWatermark lowWatermark) {
+        this.catalogService = catalogService;
+        this.lowWatermark = lowWatermark;
+    }
+
+    @Override
+    public @Nullable StorageIndexDescriptor get(int indexId) {
+        // Search for the index in the catalog history, which versions correspond to (lowWatermark, now] timestamp range.
+        int latestCatalogVersion = catalogService.latestCatalogVersion();
+
+        HybridTimestamp lowWatermarkTimestamp = lowWatermark.getLowWatermark();
```

Comment:
I think there may be a race here, as soon as we read the lwm it can immediately change and the index should already begin to be destroyed, I think this can be repaired by a consumer who will give the lwm and will not change it while you are working with it.

Reply:
Okay, let's leave a comment then.

Reply:
Fixed